Build: Prepare tooling for Gutenberg extensions in Jetpack

sirreal opened the pull request:
Adds files from Jetpack Gutenberg extensions folder (about to be merged in https://github.com/Automattic/jetpack/pull/11633) to ignore lists relevant for the packaged version of the plugin.

#### Changes proposed in this Pull Request:

* SVN Ignores
* CDN asset ignores
* Ignores for release branches

#### Testing instructions:
Together with https://github.com/Automattic/jetpack/pull/11633 (`git merge --squash --no-commit origin/add/jetpack-blocks-src`) — does the build ignore assets in `/extensions/*` but still sees `*.php` files?

- `yarn build-production`
- `php bin/build-asset-cdn-json.php`

An alternative way of testing this would be to:

- Checkout this branch. 
- Add the blocks src to: `git merge --squash --no-commit origin/add/jetpack-blocks-src`
- Comment out this line so we can build a release branch from our current branch:
  https://github.com/Automattic/jetpack/blob/e450f018a3612268e293d299a4f144948ddf3bc5/tools/build-release-branch.sh#L102
  You can apply this patch:
  https://gist.github.com/sirreal/2245987078998c2b6b9f95ba45d93d61#file-tooling-patch
- Commit the changes (`git add -A && git commit -am 'DROPME: testing changes'`)
- Run `tools/build-release-branch.sh -n`
- Pick a unique branch name.
- Let the build happen for both branches (the release branch and the built branch)
- Once the build is over, check the built branch
- It should not include any of the block's source files.
- the `.gitignore` should include the new lines.


#### Proposed changelog entry for your changes:

* None

sirreal committed:
Update release tooling for block editor extensions

"Gutenberg" block editor extensions will be moving into the repository
soon. Update related tooling accordingly to remove irrelevant source
files.

- SVN ignore for public releases
- Git release branch tooling
- CDN manifest tooling

simison commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+*.js linguist-language=JSX
```
🦆

sirreal replied:
```suggestion
```

simison replied:
https://github.com/Automattic/jetpack/pull/11639/commits/07f518ed68ec20f6b8363e2278bb8d456e2c5a47

simison commented on bin/build-asset-cdn-json.php:
```diff
@@ -9,7 +9,7 @@
 	$file = str_replace( $path, '', $file );
 	$directory = substr( $file, 0, strpos( $file, '/' ) );
-	if ( in_array( $directory, array( 'node_modules', 'tests' ) ) ) {
+	if ( in_array( $directory, array( 'node_modules', 'tests', 'extensions' ) ) ) {
```
Is this looking for `./extensions` or `./**/extensions`?

lezama replied:
I think we need to make sure it is `./extensions` right?

simison replied:
Correct, we don't want to ignore `./something/extensions`

lezama replied:
tested, works as expected, that list involves just top level paths

sirreal replied:
We're comparing the ignore list against the repo root. I've attempted to clarify things in 1915e8ba6f041b88d29e45819da9763d17a5de7e

sirreal commented on .svnignore:
```diff
@@ -42,3 +42,4 @@ docker
 bin/pre-commit-hook.js
 bin/travis_install.sh
 yarn-error.log
+extensions/**/*.{js,jsx,gif,png,css,jpg,jpeg,sass,scss}
```
```suggestion
extensions/**/*.{js,jsx,json,gif,png,css,jpg,jpeg,sass,scss}
```

simison replied:
What do you think about doing a reverse?

At least this works in bash:
```bash
ls !(*.php)
```

On second thought, that might be potentially harmful vs explicitly defining what we don't want?

kraftbj replied:
I'd rather be explicit in svnignore as it can be really surprising. Something works great in GItHub and unless someone tests that explicit feature using a beta that was shipped in SVN we could have a pretty big gap in the production release.

To help future groking, I'd love for svnignore comments (I believe `#` preceding the line should work here too) to explain why these files are ignored. Someone new to JP development may be confused why we're ignoring jpg, css, etc.

simison commented on .svnignore:
```diff
@@ -42,3 +42,4 @@ docker
 bin/pre-commit-hook.js
 bin/travis_install.sh
 yarn-error.log
+extensions/**/*.{js,jsx,json,gif,png,css,jpg,jpeg,sass,scss}
```
Copying over https://github.com/Automattic/jetpack/pull/11639#discussion_r267776127 for visibility:

> To help future groking, I'd love for svnignore comments (I believe # preceding the line should work here too) to explain why these files are ignored. Someone new to JP development may be confused why we're ignoring jpg, css, etc.

lezama replied:
suggested by @dereksmart, could we do everything except `php` instead?

simison replied:
See https://github.com/Automattic/jetpack/pull/11639#discussion_r267761123

sirreal replied:
`svnignore` _is not a real thing_. Under the hood, this is relying on shell globbing:

https://github.com/Automattic/jetpack/blob/f8078c2cd12ac508334da2fb08e37a92cf283c14/tools/build-jetpack.sh#L70-L75

We should be extremely careful and keep that in mind.

jeherve reviewed (dismissed):
I think we'll need to add the files here too:
https://github.com/Automattic/jetpack/blob/36a8543b9933b9d4a287f5f7ad7f2dc45bed1389/tools/build-release-branch.sh#L53

simison commented on .svnignore:
```diff
@@ -42,3 +42,4 @@ docker
 bin/pre-commit-hook.js
 bin/travis_install.sh
 yarn-error.log
+extensions/**/*.{js,jsx,json,gif,png,css,jpg,jpeg,sass,scss}
```
Do we need svg?

simison commented on modules/photon-cdn/jetpack-manifest.php:
```diff
@@ -1,505 +1,534 @@
-  500 => 'modules/lazy-images/js/lazy-images.js',
-  501 => 'modules/wpgroho.js',
-);
+<?php 
```
@jeherve @kraftbj do we need to upgrade this here or you'll do this during the build when packaging the release?

jeherve replied:
It's done automatically during the release process.

simison commented on bin/build-asset-cdn-json.php:
```diff
@@ -6,15 +6,24 @@
+	$top_parent_path = substr( $relative_path_to_file, 0, strpos( $relative_path_to_file, '/' ) ) ;
+	// $top_parent_path is the first folder on .
+	// empty( $top_parent_path ) is the parent directory we can any ignore the scripts in there as well.
+	if ( empty( $top_parent_path ) || in_array( $top_parent_path, array( 'node_modules', 'tests', 'tools', 'extensions', 'docker', 'bin', 'vendor', 'logs', 'docs' ) ) ) {
```
Just for clarity, this array could be its own variable further up in the file?

sirreal replied:
I restructured this and hope it's more clear now.

simison commented on bin/build-asset-cdn-json.php:
```diff
@@ -6,15 +6,24 @@
+$assets = $export;\r\n" );
+
+
+
```
nit; 3 empty lines instead of one, linters somewhere might complain :-)

simison commented on bin/build-asset-cdn-json.php:
```diff
@@ -6,15 +6,24 @@
 $export = var_export( $manifest, true );
 
-file_put_contents( $path . 'modules/photon-cdn/jetpack-manifest.php', "<?php \r\n\$assets = $export;\r\n" );
+file_put_contents( $path . 'modules/photon-cdn/jetpack-manifest.php', "<?php 
```
There's empty trailing space in `<?php `

sirreal commented on modules/photon-cdn/jetpack-manifest.php:
```diff
@@ -1,505 +1,292 @@
-  2 => 'tools/builder/transform-relative-paths.js',
-  3 => 'tools/builder/admin-css.js',
-  4 => 'bin/pre-commit-hook.js',
-  5 => 'css/jetpack-admin.min.css',
```
I'm not sure what would build things in the CSS directory, but I think they just aren't on my local machine when this list is generated. I don't believe the directory is erroneously ignored.

sirreal replied:
Oh, apparently I found them 🤷‍♂️

ockham commented on tools/build-release-branch.sh:
```diff
@@ -52,6 +52,7 @@ function create_release_gitignore {
 
     # Add custom stuff to .gitignore release
     echo "/_inc/client" >> .gitignore
+    echo "/extensions/**/*.{js,jsx,json,gif,png,css,jpg,jpeg,sass,scss,svg}" >> .gitignore
```
Maybe add a note that there's a line in `.svnignore` that needs to be synced with this (or vice versa)?

ockham replied:
eb96cbe

ockham commented on bin/build-asset-cdn-json.php:
```diff
@@ -1,20 +1,48 @@
+	// Ignore explicit ignore list.
+	foreach ( $ignore_paths as $ignore_path ) {
+		if ( 0 === strpos( $path_from_repo_root, $ignore_path ) ) {
+			continue 2;
```
TIL [`continue 2;`](http://php.net/manual/en/control-structures.continue.php)

> Note: In PHP the switch statement is considered a looping structure for the purposes of continue. continue behaves like break (when no arguments are passed). If a switch is inside a loop, continue 2 will continue with the next iteration of the outer loop.

ockham commented on bin/build-asset-cdn-json.php:
```diff
@@ -1,20 +1,48 @@
+file_put_contents( $path . 'modules/photon-cdn/jetpack-manifest.php', "<?php
+// This file is autogenerated by bin/build-asset-cdn-json.php
+
+$assets = $export;\r\n" );
```
This gives a

```
PHP Notice:  Undefined variable: assets in ~/jetpack/bin/build-asset-cdn-json.php on line 48
```

Can we suppress that somehow?

ockham replied:
E.g. using a [`HEREDOC` vor `NOWDOC`](https://stackoverflow.com/a/1848974)? (those names! :joy: )

ockham replied:
Oh, the output seems actually broken right now:

```php
<?php
// This file is autogenerated by bin/build-asset-cdn-json.php

 = array (
  0 => 'css/jetpack-icons.min.css',
  1 => 'css/jetpack-icons-rtl.css',
  2 => 'css/jetpack-idc-admin-bar.min.css',
  3 => 'css/jetpack-admin-rtl.css',
/* etc */
```

sirreal replied:
Whoops… I'll grab that…

sirreal replied:
f02923305501c38564fa55a2cab08cdff0205585

ockham commented:
> #### Testing instructions:
> Together with https://github.com/Automattic/jetpack/pull/11633 — does the build ignore assets in `/extensions/*` but still sees `*.php` files?
> 
> - `yarn build-production`
> - `php bin/build-asset-cdn-json.php`

Where do I check for presence of `*.php` files exactly? I opened `modules/photon-cdn/jetpack-manifes.php`, but that file doesn't reference any PHP files... :grimacing:

ockham reviewed (dismissed):
LGTM -- I think the [generated file](https://github.com/Automattic/jetpack/pull/11639/files#diff-979770afbf95aaea10e77dc105d32a1d) is looking good :+1:

lezama reviewed (dismissed):
come on @dereksmart

kraftbj commented:
Assigning to @jeherve for review, scheduled for today.

jeherve reviewed (changes requested):
This doesn't seem to work in my tests. When building a release branch, the files are still added to the branch, although I do see that `/extensions/**/*.{js,jsx,json,gif,png,css,jpg,jpeg,sass,scss,svg}` is properly added to `.gitignore`.

Is `{}` a supported format in `.gitignore`?

kraftbj commented:
I don't think so https://git-scm.com/docs/gitignore

jeherve commented on .svnignore:
```diff
@@ -42,3 +42,4 @@ docker
 bin/pre-commit-hook.js
 bin/travis_install.sh
 yarn-error.log
+extensions/**/*.{js,jsx,json,gif,png,css,jpg,jpeg,sass,scss,svg}
```
Should we add `md` files to that list? We probably don't need to ship the different readme in there.

sirreal replied:
I've tried to include everything here that is a frontend oriented asset, things that are clearly used to build client assets.

`.md` is a bit different, purely informative. I'm not sure whether that sort of thing should be excluded or not. If you think so I'm happy to add it to the list 🙂

jeherve replied:
I would add .md files to that list; those files won't be very useful without the rest of the block files anyway.

sirreal replied:
6a19bed ignores `.md` files in extensions

jeherve reviewed (commented):
Should we also exclude everything in the `shared` and `utils` directories inside `extensions`, regardless of the file type?

sirreal commented:
> Should we also exclude everything in the shared and utils directories inside extensions, regardless of the file type?

I can imagine files we expect to be present in the released plugin winding up in these directories, but I really don't have a strong opinion one way or the other. Again, if you do have a strong opinion happy to make the change 👍

sirreal commented:
I've updated the gitignore and svnignore after manually testing the commands against my repo. I believe this is working well now.

simison commented on tools/build-release-branch.sh:
```diff
@@ -52,6 +52,12 @@ function create_release_gitignore {
+    printf "%s\n" "/extensions/**/*."{"js","jsx","json","gif","png","css","jpg","jpeg","sass","scss","svg"} >> .gitignore
 
     # Remove old .gitignore
     rm .gitignore-tmp
```
Was about to comment about spaces as indentation on your new lines but realised this file is already a mixture of tabs/spaces.

simison commented on tools/build-release-branch.sh:
```diff
@@ -52,6 +52,12 @@ function create_release_gitignore {
+    # .gitignore will not respect the {brace,variations} so we'll produce multiple lines.
+    #
+    #                       ↓ this brace is unquoted. Important!
+    printf "%s\n" "/extensions/**/*."{"js","jsx","json","gif","png","css","jpg","jpeg","sass","scss","svg"} >> .gitignore
```
Was about to comment about spaces as an indentation on your new lines but realised this file is already a mixture of tabs/spaces... so we can ignore this.

sirreal replied:
Fixed indentaion and made this file more consistent in general in 99fbf91

jeherve reviewed (dismissed):
I am afraid it still does not work. The `.gitignore` is properly updated, but the files still get committed to the built branch (see the instructions I posted in the PR description).

simison commented on tools/build-release-branch.sh:
```diff
@@ -52,6 +52,12 @@ function create_release_gitignore {
+    # .gitignore will not respect the {brace,variations} so we'll produce multiple lines.
+    #
+    #                       ↓ this brace is unquoted. Important!
+    printf "%s\n" "/extensions/**/*."{"js","jsx","json","gif","png","css","jpg","jpeg","sass","scss","svg"} >> .gitignore
```
Alternatively, we could simplify this to just echoing multiple lines or looping a simple array. That might help to catch any issues and help people to expand this if need.

sirreal replied:
Good call. Simplified in e450f018a3612268e293d299a4f144948ddf3bc5

sirreal committed:
Add md to ignored files

Alphabetize extensions ignores

sirreal committed:
Cleanup space and block open style

Keep block openers on the same line:

if [ cond ]; then
while [ cond ]; do

sirreal committed:
Simplify .gitignore output

sirreal commented on tools/build-release-branch.sh:
```diff
@@ -16,117 +17,130 @@
+	echo "/extensions/**/*.png" >> .gitignore
+	echo "/extensions/**/*.sass" >> .gitignore
+	echo "/extensions/**/*.scss" >> .gitignore
+	echo "/extensions/**/*.svg" >> .gitignore
```
Shall we add `/docker/` here as well?

lezama replied:
yess

sirreal committed:
Remove Jest snapshots from releases

sirreal commented:
I've rebased and cleaned up this branch. I've removed the accidental merge from #11633 and dropped any changes to the generated manifest file.

I suspect the changes from  f53c4c6ac1682dac76090311be10258088dbe38b (props @lezama) will address the most recent review:

> I am afraid it still does not work. The .gitignore is properly updated, but the files still get committed to the built branch (see the instructions I posted in the PR description).

Here's what my build `extensions` directory looks like after testing:

```
extensions/
└── blocks
    ├── business-hours
    │   └── business-hours.php
    ├── contact-info
    │   ├── class-jetpack-contact-info-block.php
    │   └── contact-info.php
    ├── gif
    │   └── gif.php
    ├── mailchimp
    │   └── mailchimp.php
    ├── map
    │   └── map.php
    ├── markdown
    │   └── markdown.php
    ├── repeat-visitor
    │   └── repeat-visitor.php
    ├── slideshow
    │   └── slideshow.php
    ├── tiled-gallery
    │   ├── layout
    │   │   └── mosaic
    │   │       └── test
    │   └── tiled-gallery.php
    ├── vr
    │   └── vr.php
    └── wordads
        └── wordads.php

15 directories, 12 files
```

sirreal commented on tools/build-release-branch.sh:
```diff
@@ -16,117 +17,131 @@
+	done < ".gitignore-tmp"
+
+	# Add custom stuff to .gitignore release
+	echo "/_inc/client" >> .gitignore
```
```suggestion
	echo "/_inc/client" >> .gitignore
	echo "/docker/" >> .gitignore
```

sirreal committed:
Add docker to release branch gitignore

ockham commented on tools/build-release-branch.sh:
```diff
@@ -16,117 +17,133 @@
+	echo "/extensions/**/*.sass" >> .gitignore
+	echo "/extensions/**/*.scss" >> .gitignore
+	echo "/extensions/**/*.svg" >> .gitignore
+	echo "__snapshots__/" >> .gitignore
```
Late to the `.gitignore` party, but would a (semi-)whitelist-based approach (i.e. using negation) work? Something like

```
echo "/extensions/" >> .gitignore
echo "!/extensions/**/*.php" >> .gitignore
```

See e.g. https://www.atlassian.com/git/tutorials/saving-changes/gitignore#git-ignore-patterns

sirreal replied:
That would likely work, but would be harder to keep in sync with the `.svnignore` which is really a list of files to pass to `rm -fr`.

lezama replied:
why we need a `.svnignore` again?

sirreal replied:
It's used in several places around the build/deploy tooling:

https://github.com/Automattic/jetpack/blob/f8078c2cd12ac508334da2fb08e37a92cf283c14/tools/build-jetpack.sh#L70-L75
https://github.com/Automattic/jetpack/blob/58ec3030c9d79e10ace25574cddfaf6d2f50fe6a/tools/build-release-branch.sh#L228-L235
https://github.com/Automattic/jetpack/blob/f8078c2cd12ac508334da2fb08e37a92cf283c14/tools/deploy-to-master-stable-branch.sh#L67-L73
https://github.com/Automattic/jetpack/blob/cadfa61c28b6ed30015313748feb959dfea60235/tools/deploy-to-svn.sh#L72-L74

ockham replied:
Not for this PR, but I wonder if we could de-dupe sources of truth (i.e. use `.gitifnore` where we're now using `.svnignore`) in the long run... :thinking:

ockham commented on tools/build-release-branch.sh:
```diff
@@ -16,117 +17,133 @@
+
+			git rm -r --cached .
+			git add .
+			git commit -m ".gitignore cleanup"
```
Maybe make the commit msg more explicit here? Like `Remove files based on .gitignore contents`?

ockham commented on tools/build-release-branch.sh:
```diff
@@ -16,117 +17,133 @@
+			modify_svnignore
+
+			git rm -r --cached .
+			git add .
```
Do we need this line? Not seeing anything added by that commit :thinking:

sirreal replied:
I think the logic here is
- Remove everything
- Add everything (respecting our new .gitignore)

This is the way we're able to filter out versioned files that are newly added to the gitignore.

ockham reviewed (dismissed):
Left a few minor (non-blocking) suggestions. Overall, this is looking great and working well AFAICT!

:ship: at will!

jeherve commented on tools/build-release-branch.sh:
```diff
@@ -16,117 +17,133 @@
+			modify_svnignore
+
+			git rm -r --cached .
+			git add .
```
I don't think this is ideal. While it works, we end up committing the updated Jetpack dashboard build on this step instead of at the next one. It also has the potential to be quite dangerous. For example, if you are using a custom version of the Gutenberg blocks locally (after running `npx lerna run build --stream --scope='@automattic/jetpack-blocks'` for example), those changes will be committed to the release branch. We don't want that to happen, as the blocks should always be the ones that have been tagged in npm for now.

sirreal replied:
Would you prefer something like the following, mirroring the lines above?

```suggestion
git rm -fg \
	extensions/**/*.css \
	extensions/**/*.gif \
	extensions/**/*.jpeg \
	extensions/**/*.jpg \
	extensions/**/*.js \
	extensions/**/*.json \
	extensions/**/*.jsx \
	extensions/**/*.md \
	extensions/**/*.png \
	extensions/**/*.sass \
	extensions/**/*.scss \
	extensions/**/*.svg \
	**/__snapshots__
# no `git add .` required
```

jeherve replied:
Yeah, maybe removing them like that would be better, since it wouldn't impact any other code by mistake?

jeherve replied:
It seems it helped, but I now see some .js files still committed to the release branch:

![image](https://user-images.githubusercontent.com/426388/54930547-bcaff100-4f17-11e9-90fa-a4405e7023d7.png)

Did it work for you?

kraftbj replied:
We had a long chat about this and we're going to accept the source files in the built branches for now as it matches what happens with `_inc/client` right now. We're going to offload stripping these to the deploy-to-svn and .svnignore convention.

Fixing this up in the git built branch is a later thing to tackle.

jeherve reviewed (dismissed):
While this appears to work for now, I have some reservations about the latest changes; they can potentially be dangerous in that you can end up committing things you did not mean to.

sirreal commented on tools/build-release-branch.sh:
```diff
@@ -16,117 +17,133 @@
+
+			git rm -r --cached .
+			git add .
+			git commit -m ".gitignore cleanup"
```
This sound good?
```suggestion
			git commit -m 'Remove undesired files from new release branch

Some files from "working" branches will not be included in the released
version of the plugin. This commit ignores and removes known files from
the release branch.'
```

sirreal committed:
Remove undesired files explicitly

sirreal committed:
Improve gitignore message

sirreal commented:
I've added fdaeab9 and 327218f to address the latest review.

sirreal committed:
Revert source files removal

sirreal commented:
After a call with @kraftbj and @dereksmart representing @Automattic/jetpack-crew, I've rolled back the changes that were intended to remove many files from extensions in release branches.

Part of the reasoning is that other unbuilt source is already included in release branches so including source in beta branches is not a regression and not something that needs to be addressed in this PR.

Thanks everyone, reviews appreciated 🙇

ockham reviewed (approved):
Seems to work as described and expected :+1:

kraftbj reviewed (approved):
I'm cautious about the additions in `.svnignore` and wanting to make sure they work as expected, so this will be an area of heavy testing in the next couple of days.
